core/services/pipeline/runner.go

```diff
@@ -48,7 +49,9 @@ type Runner interface {
 	// ExecuteAndInsertFinishedRun executes a new run in-memory according to a spec, persists and saves the results.
 	// It is a combination of ExecuteRun and InsertFinishedRun.
 	// Note that the spec MUST have a DOT graph for this to work.
-	ExecuteAndInsertFinishedRun(ctx context.Context, spec Spec, vars Vars, l logger.Logger, saveSuccessfulTaskRuns bool) (runID int64, finalResult FinalResult, err error)
+	ExecuteAndInsertFinishedRun(ctx context.Context, spec Spec, vars Vars, l logger.Logger, saveSuccessfulTaskRuns bool) (runID int64, results TaskRunResults, err error)
+	// ExecuteAndInsertFinishedRunWithSpec works just like ExecuteAndInsertFinishedRun but also stores the spec in the DB.
+	ExecuteAndInsertFinishedRunWithSpec(ctx context.Context, spec Spec, vars Vars, l logger.Logger, saveSuccessfulTaskRuns bool) (runID int64, results TaskRunResults, err error)
```

Comment:
@vyzaldysanchez Can we extend ExecuteAndInsertFinishedRun to create the spec if it doesn't have an associated ID rather than adding a new method? This will lead to a deeper interface which is usually preferable to the wider one you're introducing here.

Reply:
Sure thing.

Reply:
This is now addressed.